File: LeetCode-c++/216_combinationSum3.cpp
#include<iostream>
#include<vector>
using namespace std;

class Solution
{
    public:
    vector<vector<int>> result;
    vector<vector<int>> combinationSum3(int k, int n)
    {
        vector<int> path;
        backtrack(1, 0, k, n, path);
        return result;
    }
    void backtrack(int start, int sum, int k, int n, vector<int>& path)
    {
        for(int i=start; i<10; ++i)
        {
            path.push_back(i);
            sum += i;
            if(path.size()>k)
            {
                path.pop_back();
                sum -= i;
                return;
            }
            if(sum>=n){
                if(sum==n && path.size()==k)
                    result.push_back(path);
                path.pop_back();
                sum -= i;
                return;
            }
            backtrack(i+1, sum, k, n, path);
            path.pop_back();
            sum -= i;
        }
    }
};

int main()
{
    int k = 3;
    int n = 7;
    Solution solution = Solution();
    vector<vector<int>> res = solution.combinationSum3(k, n);
    for(vector<int> x: res)
    {
        cout << '[';
        for(int i: x)
        {
            cout << i << ' ';
        }
        cout << ']' <<endl;
    }
    system("pause");
    return 0;
}

File: LeetCode-py/216_combinationSum3.py
class Solution:
    @classmethod
    def combinationSum3(self, k, n):
        result = []
        def backtrack(start, sum, path):
            for i in range(start, 10):
                sum += i
                if(len(path)>=k):
                    return
                if(sum==n and len(path)==k-1):
                    result.append(path+[i])
                    sum -= i
                    return
                if(sum>n):
                    sum -= i
                    return
                backtrack(i+1, sum, path+[i])
                sum -= i

        backtrack(1, 0, [])
        return result
if __name__ == "__main__":
    k = 3
    n = 9
    print(Solution.combinationSum3(k, n))